dask/array/reductions.py

```diff
@@ -175,19 +182,34 @@ def _tree_reduce(x, aggregate, axis, keepdims, dtype, split_every=None,
         if i in split_every and split_every[i] != 1:
             depth = int(builtins.max(depth, ceil(log(n, split_every[i]))))
     func = partial(combine or aggregate, axis=axis, keepdims=True)
+    meta_func = None if reduced_meta is None else func
     if concatenate:
         func = compose(func, partial(_concatenate2, axes=axis))
     for i in range(depth - 1):
         x = partial_reduce(func, x, split_every, True, dtype=dtype,
-                           name=(name or funcname(combine or aggregate)) + '-partial')
+                           name=(name or funcname(combine or aggregate)) + '-partial',
+                           meta_func=meta_func, reduced_meta=reduced_meta)
     func = partial(aggregate, axis=axis, keepdims=keepdims)
+    meta_func = None if reduced_meta is None else func
     if concatenate:
         func = compose(func, partial(_concatenate2, axes=axis))
     return partial_reduce(func, x, split_every, keepdims=keepdims, dtype=dtype,
-                          name=(name or funcname(aggregate)) + '-aggregate')
+                          name=(name or funcname(aggregate)) + '-aggregate',
+                          meta_func=meta_func, reduced_meta=reduced_meta)
+
 
+def try_combinations(func, e, x, combinations):
+    if len(combinations) == 1:
+        return func(x, **combinations[0])
+
+    try:
+        return func(x, **combinations[0])
+    except e:
+        return try_combinations(func, e, x, combinations[1:])
```

Comment:
Can I ask for information about this?  Why did we need this?

Reply:
Or, maybe more directly, the fact that we need this is somewhat concerning.  I'm curious about what the problem is so that maybe someone can find another approach.

Reply:
I don't like this either, but what happens here is that in `partial_reduce()` I'm calling `func()` (or `meta_func()` which doesn't `_concatenate2()`), and depending on the function in question, it may accept different arguments. In particular the `meta` argument had to be added because functions chunk/combine/aggregate functions may return dicts, which aren't valid arrays to reduce.

Reply:
Yeah I agree with both of you, we should find another approach here if possible. Do you have a couple of examples where this is needed so we can formulate and try out some alternatives?

Reply:
Alright, I pushed commit https://github.com/dask/dask/pull/4543/commits/ce7e5ddf852087cee5fc75effdb99c868afbd772 to get rid of (part of) that. We still need some exception handling for some of the reduction functions (e.g., `mean_chunk`, `mean_combine`, etc.) for which I added a `meta=` keyword argument to prevent them from returning `dict`s that we don't want for `_meta` construction, but most of them won't have such keyword. The second case is for `ValueError` exception, like `ufunc`s that can't be executed on empty arrays.

I now honestly don't know if I had a legitimate case where I've failure due to `dtype=` and `keepdims=`. The two possibilities I see now for that are:

1. I had some bug in another part of the code during my development process; or
1. I'm really going insane.

Please let me know if the solution on the aforementioned commit seems acceptable. But IMHO, it seems acceptable.